core/src/main/java/org/infinispan/encoding/ProtostreamTranscoder.java

```diff
@@ -113,19 +114,27 @@ public Object transcode(Object content, MediaType contentType, MediaType destina
    }
 
    private byte[] marshall(Object decoded, MediaType destinationType) throws IOException {
-      Optional<String> wrappedParam = destinationType.getParameter("wrapped");
-      if (!wrappedParam.isPresent() || !wrappedParam.get().equals("false"))
-         return ProtobufUtil.toWrappedByteArray(ctx(), decoded);
-      return ProtobufUtil.toByteArray(ctx(), decoded);
+      try {
+         Optional<String> wrappedParam = destinationType.getParameter("wrapped");
+         if (!wrappedParam.isPresent() || !wrappedParam.get().equals("false"))
+            return ProtobufUtil.toWrappedByteArray(ctx(), decoded);
+         return ProtobufUtil.toByteArray(ctx(), decoded);
+      } catch (IllegalArgumentException iae) {
+         throw new MarshallingException(iae.getMessage());
+      }
    }
 
    private Object unmarshall(byte[] bytes, MediaType destinationType) throws IOException {
-      String type = destinationType.getClassType();
-      if (type == null) {
-         return ProtobufUtil.fromWrappedByteArray(ctx(), bytes);
+      try {
+         String type = destinationType.getClassType();
+         if (type == null) {
```

Comment:
When producing protobuf from ```Something```, it's possible to serialize to the protobuf wire format directly or to serialize using a message wrapper containing the protobuf payload, aka  ```org.infinispan.protostream.WrappedMessage```

This is used in RemoteQuery, that needs to serialize to those two formats (user payload and query structure itself)

AFAICT, when umarshalling, it will automatically work :smile:

I will add some test/doc to this

Reply:
:+1: 

I don't doubt this feature is needed, I just wanted an explanation of how the `wrapped` and `type` parameters work together.

Reply:
Hope it makes sense now?  ;)